src/ex_cmds.c

```diff
@@ -4480,8 +4480,18 @@ ex_substitute(exarg_T *eap)
 		    nmatch_tl += nmatch - 1;
 		}
 		copy_len = regmatch.startpos[0].col - copycol;
-		needed_len = copy_len + ((unsigned)STRLEN(p1)
+
+		/*
+		 * Make sure needed_len always greater than copy_len
+		 */
+		if((unsigned)STRLEN(p1) == 0)	{
```

Comment:
STRLEN is O(n). To check whether it's an empty string, you can do it in O(1) with:
```
if (*p1 == NUL)
```
Although the compiler might optimize it anyway.